### PR TITLE
Add new gflag for a total time limit.

### PR DESCRIPTION
**Pull request by ajf101**
This creates a new flag, `-g` followed by seconds, which is the maximum length of time to run tcpdump for.

### CODE REVIEW

File: debian/patches/gflag-time-limit.diff [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
++	}
+ }
+ 
+ /*
+@@ -3161,6 +3209,22 @@
+ }
+ #endif /* _WIN32 */
+ 
++/*
++ * Called once each second if gflag is set
```
Comment (jmb202) [inline debian/patches/gflag-time-limit.diff]: Not on unix it isn't -- you've scheduled the signal to be fired after gflag seconds.

File: debian/patches/gflag-time-limit.diff [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
++	/* Get the current time */
++	if ((t = time(NULL)) == (time_t)-1) {
++		error("%s: can't get current_time: %s",
++			__func__, pcap_strerror(errno));
++	}
++
++	/* If the time hits the expected end time, exit */
++	if (t >= gflag_time) {
++		gflag_time_exceeded();
++	}
```
Comment (jmb202) [inline debian/patches/gflag-time-limit.diff]: Why is this needed? The timer should fire and handle this already, no?

File: debian/patches/gflag-time-limit.diff [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+ 			infile = optarg;
+ 			break;
+ 
++		case 'g':
++			gflag = atoi(optarg);
++			if (gflag < 1)
++				error("invalid number of seconds %s", optarg);
++
++			/* Grab the current time for time limit. */
++			if ((gflag_time = time(NULL)) == (time_t)-1) {
```
Comment (jmb202) [inline debian/patches/gflag-time-limit.diff:39]: Do we care that time can go backwards if NTP adjusts it aggressively?

File: debian/patches/gflag-time-limit.diff (hunk 2 above)
Comment (ajf101) [reply debian/patches/gflag-time-limit.diff]: This is if SIGALRM is already in use for the verbose logging mode.

File: debian/patches/gflag-time-limit.diff (hunk 3 above)
Comment (ajf101) [reply debian/patches/gflag-time-limit.diff:39]: It is a plausible scenario, however not one that I think needs worrying about: The existing periodic config (-G) would also be affected here.

Comment (jmb202) [reply debian/patches/gflag-time-limit.diff:39]: Fine. Meh.

File: debian/patches/gflag-time-limit.diff (hunk 2 above)
Comment (jmb202) [reply debian/patches/gflag-time-limit.diff]: Er. Right. But that's not gonna work if no gflag is specified? (gflag_time will be zero in that case, as it's statically allocated)

Comment (jmb202) [reply debian/patches/gflag-time-limit.diff]: It's somewhat confusing having two different paths here, tbh -- can they not be unified such that there is a single way that the timer can be set up, and thus a single entry point from the timer that handles either case as appropriate?

Comment (ajf101) [reply debian/patches/gflag-time-limit.diff]: Thanks for spotting the lack of the `if gflag` test - fixed.

The timer has two very different paths:
- Every second interval timer for printing stats
- Single timer for the gflag value

Both end up in `gflag_time_exceeded()` to actually do the exit.

But I don't think we can get much other code sharing here unless we handled the second case by an interval timer as well. Which didn't seem efficient; if we don't need any other periodic call then going straight to a single `alarm()` call seemed the most efficient way of handling it. The time checking only comes into play when SIGALARM is coopted by the interval timer.

File: debian/patches/gflag-time-limit.diff [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
++static int gflag;			/* exit after this many seconds */
++static time_t gflag_time;		/* time_t to exit at */
+ static int Gflag;			/* rotate dump files after this many seconds */
+ static int Gflag_count;			/* number of files created with Gflag rotation */
+ static time_t Gflag_time;		/* The last time_t the dump file was rotated. */
+@@ -228,6 +230,7 @@
+ #ifdef HAVE_PCAP_DUMP_FLUSH
+ static int Uflag;			/* "unbuffered" output of dump files */
+ #endif
++static int vflag_print;     /* indicates if verbose stats should be dumped */
```
Comment (jmb202) [inline debian/patches/gflag-time-limit.diff]: Whitespace

File: debian/patches/gflag-time-limit.diff [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
++	if (gflag > 0 || vflag_print) {
+ 		/*
+ 		 * When capturing to a file, if "--print" wasn't specified,
+ 		 *"-v" means tcpdump should, once per second,
+@@ -2477,6 +2496,8 @@
+ 		 * Except when reading from a file, because -r, -w and -v
+ 		 * together used to make a corner case, in which pcap_loop()
+ 		 * errored due to EINTR (see GH #155 for details).
++         *
++         * Also, set up this timer if the gflag is set.
```
Comment (jmb202) [inline debian/patches/gflag-time-limit.diff]: Ditto

File: debian/patches/gflag-time-limit.diff [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
++		if ((t = time(NULL)) == (time_t)-1) {
++			error("%s: can't get current_time: %s",
++				__func__, pcap_strerror(errno));
++		}
++
++		/* If the time hits the expected end time, exit */
++		if (t >= gflag_time) {
++            (void)fprintf(stderr, "Time limit reached: %d seconds\n", gflag);
++            info(1);
++            exit_tcpdump(0);
```
Comment (jmb202) [inline debian/patches/gflag-time-limit.diff]: Ditto.

**Review by jmb202 — APPROVED** [review]
3 minor whitespace nits, otherwise lgtm.

File: debian/changelog [hunk 7]
```diff
@@ -1,3 +1,9 @@
+tcpdump (4.99.3-1+pexip24u5) pexip-bookworm; urgency=medium
```
Comment (jmb202) [inline debian/changelog]: `s/pexip-bookworm/pexip/`
